### PR TITLE
build: add Lighthouse CI integration (resolves #267)

### LINKED ISSUE #267: Integrate Lighthouse CI into Github Actions workflow
## Is your feature request related to a problem? Please describe.

We've been using Lighthouse through Chrome's dev tools to generate performance reports for the website. In particular for the integration of UIO. For example: #230 

## Describe the solution you'd like

[Lighthouse CI](https://github.com/GoogleChrome/lighthouse-ci/blob/master/docs/getting-started.md) provides CI integration for Github Actions. By adding this, it will allow us to integrate it into the existing CI infrastructure and have performance tests run on PRs and other Pushes.

## Describe alternatives you've considered

There are other ways to use Lighthouse, e.g through the Chrome Dev tools, [Firefox Browser Extension](https://addons.mozilla.org/en-US/firefox/addon/google-lighthouse/), [Chrome Browser Extension](https://chrome.google.com/webstore/detail/lighthouse/blipmdconlkpinefehnmjammfjpmpbjk), [Lighthouse Node Module](https://www.npmjs.com/package/lighthouse),  and Google services such as [web.dev/measure](https://web.dev/measure/) and [PageSpeed Insights](https://developers.google.com/speed/pagespeed/insights/). This does not preclude anyone from using those options as well, and provides seamless integration into the PR and Push process.

I have not explored other performance analysis tools, as it appears the the project has already settled on Lighthouse. However, a further investigation into other automated tools could be useful.

### PR DESCRIPTION
**Pull request by jobara**
* [x] This pull request has been linted by running `npm run lint` without errors
* [x] This pull request has been tested by running `npm run start` and reviewing affected routes
* [x] This pull request has been built by running `npm run build` without errors
* [x] This isn't a duplicate of an existing pull request

## Description

Adds [Lighthouse CI](https://github.com/GoogleChrome/lighthouse-ci/blob/master/docs/getting-started.md) integration.

## Steps to test

* Review the Lighthouse build results
* Review PR status checks for Lighthouse results

## Related issues

Resolves #267

### CODE REVIEW

**Comment by jobara** [comment]
@cindyli and @TeddTech could you please take a look at this PR. I have Desktop and Mobile runs, have removed throttling, removed the PWA category, and set thresholds for asserting the results for the categories. You should be able to view the results for failures through annotations on the build, which will also link to those results uploaded to the cloud. You can also download the results from the artifacts.

**Comment by cindyli** [comment]
@jobara this is super helpful. Some minScore assertion failures already help to capture some issues we didn't noticed before. I think @TeddTech or I should go through those pages individually, figure out causes and fix ones that are valid.

The minScore failures on SEO are likely because of the blocking directive on servers, which we cannot do much. We probably should adjust the minScore for it.

**Comment by TeddTech** [comment]
> @jobara this is super helpful. Some minScore assertion failures already help to capture some issues we didn't noticed before. I think @TeddTech or I should go through those pages individually, figure out causes and fix ones that are valid.
> 
> The minScore failures on SEO are likely because of the blocking directive on servers, which we cannot do much. We probably should adjust the minScore for it.

I agree the minScore for SEO can be lower

**Comment by TeddTech** [comment]
> @jobara this is super helpful. Some minScore assertion failures already help to capture some issues we didn't noticed before. I think @TeddTech or I should go through those pages individually, figure out causes and fix ones that are valid.
> 
> The minScore failures on SEO are likely because of the blocking directive on servers, which we cannot do much. We probably should adjust the minScore for it.

The outstanding failures for this build seem to be all SEO or accessibility issues.

SEO failures can be fixed by lowering its `minScore` threshold since there's not much else we can do because the failures are caused by the blocking directive on servers as @cindyli mentioned. I would suggest a `minScore` anywhere from 85-88.

The following accessibility failures are due to the webpage layout and should be brought up at tomorrow's check-in with the designers present. Below are single examples of the two types of accessibility failures seen in the build.

1. I believe this accessibility failure is due to there being [`h2` elements in the footer](https://web.dev/heading-order/?utm_source=lighthouse&utm_medium=cli#how-to-fix-poorly-structured-headings).
Screen Shot 2020-06-24 at 9 42 45 AM

2. I believe this accessibility failure is due to the choice of [foreground and background colour](https://web.dev/color-contrast/?utm_source=lighthouse&utm_medium=cli#how-to-ensure-text-has-sufficient-color-contrast).

Screen Shot 2020-06-24 at 9 46 49 AM

**Comment by jobara** [comment]
@cindyli and @TeddTech, thanks for the feedback. Instead of lowering the SEO score I just skipped the [is-crawlable](https://web.dev/is-crawable/?utm_source=lighthouse&utm_medium=cli) audit. Netlify sets the `X-Robots-Tag: noindex` response header for deploy previews, which is causing the related audit to fail.

**Comment by cindyli** [comment]
>     1. I believe this accessibility failure is due to there being [`h2` elements in the footer](https://web.dev/heading-order/?utm_source=lighthouse&utm_medium=cli#how-to-fix-poorly-structured-headings).
> 
This is because there are elements on the page using h3 after h1 and missing h2 in between. I've created the issue https://github.com/inclusive-design/wecount.inclusivedesign.ca/issues/309 to address it.

> 
>     1. I believe this accessibility failure is due to the choice of [foreground and background colour](https://web.dev/color-contrast/?utm_source=lighthouse&utm_medium=cli#how-to-ensure-text-has-sufficient-color-contrast).
> 
I've sent an email to designers to clarify contrast ratios.

**Comment by jobara** [comment]
@cindyli and @TeddTech I've updated to the latest version of dev. Also I found https://googlechrome.github.io/lighthouse-ci/viewer/ which lets you compare the results from two runs. Not sure it's useful at the moment, but something else that can be used. I also found a tool called speedlify https://www.zachleat.com/web/speedlify/ that can be used to track the live site over time.

**Comment by cindyli** [comment]
@jobara, now the lighthouse CI failures are all caused by low contrast ratios of some links. I've filed #331 for it. While we wait for the UI design for #331, do you think if we can go ahead to merge this pull request? Before the merge, off the top of my head there are two things to take care:

1. Lower the `minScore` for accessibility to 0.98;
2. Add cheks for two new pages: [learn](https://dev--wecount.netlify.app/learn/), [initiatives](https://dev--wecount.netlify.app/initiatives/).

**Comment by jobara** [comment]
> @jobara, now the lighthouse CI failures are all caused by low contrast ratios of some links. I've filed #331 for it. While we wait for the UI design for #331, do you think if we can go ahead to merge this pull request? Before the merge, off the top of my head these need to be taken care:
> 
> 1. Update to the latest dev;
> 2. Lower the `minScore` for accessibility to 0.98;
> 3. Add cheks for two new pages: [learn](https://dev--wecount.netlify.app/learn/), [initiatives](https://dev--wecount.netlify.app/initiatives/).

@cindyli I've addressed point 1 and 3 from above. I haven't lowered the `minScore` for accessibility checks just yet. I wonder if we really want to do that. I can see that not checking, lowering the threshold, and ignoring failing builds are all less than idea. Is there a timeframe for when the last bits will get in?

**Comment by jobara** [comment]
Chrome is up to v84 now and includes Lighthouse 6.0.0; which is the same version in the CI updates in this PR.

**Comment by jobara** [comment]
Also of note is how the scoring has changes between Lighthouse 5 and 6. See: https://web.dev/performance-scoring/

**Comment by cindyli** [comment]
> I haven't lowered the `minScore` for accessibility checks just yet. I wonder if we really want to do that. I can see that not checking, lowering the threshold, and ignoring failing builds are all less than idea. Is there a timeframe for when the last bits will get in?

I agree. I will talk to designers again. Hopefully to fix the contrast ratio issue before the 2.2. release.

**Comment by cindyli** [comment]
@jobara, could you merge in the "dev" branch? With the fix for #331, hope all lighthouse checks will pass. Thanks.

File: .editorconfig [hunk 1]
```diff
@@ -11,3 +11,7 @@ insert_final_newline = true
 
 [*.md]
 trim_trailing_whitespace = false
+
+[*.yml]
+indent_style = space
```
Comment (cindyli) [inline .editorconfig:16]: The indent style in WeCount is to use "tab" as it provides flexibility to map to desired number of spaces. Do you think if yml files can use it too?

Comment (jobara) [reply .editorconfig:16]: @cindyli tabs are not valid in YAML files. https://yaml.org/faq.html

**Comment by cindyli** [comment]
[The latest lighthouse CI failure](https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1595861188950-84493.report.html#best-practices) reports the image on [https://dev--wecount.netlify.app/initiatives/](https://dev--wecount.netlify.app/initiatives/) is [displayed with incorrect aspect ratio](https://web.dev/image-aspect-ratio/). However, looking at 3 sizes on the report, their calculated aspect ratios are:
* Displayed size: 306 X 160; aspect ratio: 1.9125
* Actual size: 360 X 188; aspect ratio: 1.9148
* Expected size: 918 X 480; aspect ratio: 1.9125

The difference between aspect ratios is nowhere near 5%.

I've filed [a bug report](https://github.com/GoogleChrome/web.dev/issues/3577) with Google Chrome web.dev.

**Comment by cindyli** [comment]
According to the reply to [my issue report](https://github.com/GoogleChrome/web.dev/issues/3577) with Google Chrome web.dev, the lighthouse report on "displayed with incorrect aspect ratio" _is not complaining about the aspect ratio it's complaining about the total resolution (including DPR), but it linked to the incorrect documentation_.

Moreover, the improvement for calculating the total resolution was fixed in 6.1.0 ([#10801](https://github.com/GoogleChrome/lighthouse/pull/10801)) and will be in Chrome shortly.

**Comment by jobara** [comment]
> According to the reply to [my issue report](https://github.com/GoogleChrome/web.dev/issues/3577) with Google Chrome web.dev, the lighthouse report on "displayed with incorrect aspect ratio" _is not complaining about the aspect ratio it's complaining about the total resolution (including DPR), but it linked to the incorrect documentation_.
> 
> Moreover, the improvement for calculating the total resolution was fixed in 6.1.0 ([#10801](https://github.com/GoogleChrome/lighthouse/pull/10801)) and will be in Chrome shortly.

@cindyli thanks for looking into this. Unfortunately they haven't updated Lighthouse CI yet, it's still pointing at [v6.0.0](https://github.com/GoogleChrome/lighthouse-ci/blob/b7a906d282043ad385081cf1753157b0ec2aa11b/packages/cli/package.json#L22). I'm guessing once they update and create a new release, our CI should just pick it up. In regards to total resolution, it looks like it should be pointing at [Serve responsive images](https://web.dev/serve-responsive-images/). It's not clear if their calculation adjustments will remove the issue, or if there is still more work needed on our part.

**Comment by greatislander** [comment]
I'm closing this as the Netlify Lighthouse plugin can be used instead.

**Pull request closed without merging**